drivers/cache/cache_andes.c

```diff
@@ -0,0 +1,737 @@
+		nds_l1d_cache_range(addr, size, K_CACHE_INVD);
+#endif /* CONFIG_L2C_INCLUSIVE_POLICY */
+	} else {
+		nds_l1d_cache_range(addr, size, K_CACHE_WB_INVD);
+	}
+#else
+	nds_l1d_cache_range(addr, size, K_CACHE_WB_INVD);
+#endif /* L2C_EXISTS */
+
+	k_spin_unlock(&lock, key);
+
+	return 0;
+}
+
+int cache_instr_flush_all(void)
+{
+	return -ENOTSUP;
+}
+
+int cache_instr_flush_and_invd_all(void)
+{
+	return -ENOTSUP;
+}
+
+int cache_instr_flush_range(void *addr, size_t size)
+{
+	ARG_UNUSED(addr);
+	ARG_UNUSED(size);
+
+	return -ENOTSUP;
+}
+
+int cache_instr_flush_and_invd_range(void *addr, size_t size)
+{
+	ARG_UNUSED(addr);
+	ARG_UNUSED(size);
+
+	return -ENOTSUP;
+}
+
+#ifdef CONFIG_DCACHE_LINE_SIZE_DETECT
+size_t cache_data_line_size_get(void)
+{
+	return cache_cfg.data_line_size;
+}
+#endif /* CONFIG_DCACHE_LINE_SIZE_DETECT */
+
+#ifdef CONFIG_ICACHE_LINE_SIZE_DETECT
+size_t cache_instr_line_size_get(void)
+{
+	return cache_cfg.instr_line_size;
+}
+#endif /* CONFIG_ICACHE_LINE_SIZE_DETECT */
+
+static int andes_cache_init(void)
+{
+	unsigned long line_size, ret = 0;
+
+#ifdef CONFIG_ICACHE
+	line_size = (csr_read(NDS_MICM_CFG) >> 6) & BIT_MASK(3);
+
+	if (line_size == 0) {
+		LOG_ERR("Platform doesn't support I-cache, "
+			"please disable CONFIG_ICACHE");
+	}
+#ifdef CONFIG_ICACHE_LINE_SIZE_DETECT
+	/* Icache line size */
+	if (line_size <= 5) {
+		cache_cfg.instr_line_size = 1 << (line_size + 2);
+	} else {
+		LOG_ERR("Unknown line size of I-cache");
+	}
+#elif (CONFIG_ICACHE_LINE_SIZE != 0)
+	cache_cfg.instr_line_size = CONFIG_ICACHE_LINE_SIZE;
+#elif DT_NODE_HAS_PROP(DT_PATH(cpus, cpu_0), i_cache_line_size)
+	cache_cfg.instr_line_size =
+		DT_PROP(DT_PATH(cpus, cpu_0), "i_cache_line_size");
+#else
+	LOG_ERR("Please specific the i-cache-line-size "
+		"CPU0 property of the DT");
+#endif /* CONFIG_ICACHE_LINE_SIZE_DETECT */
+#endif /* CONFIG_ICACHE */
+
+#ifdef CONFIG_DCACHE
+	line_size = (csr_read(NDS_MDCM_CFG) >> 6) & BIT_MASK(3);
+	if (line_size == 0) {
+		LOG_ERR("Platform doesn't support D-cache, "
+			"please disable CONFIG_DCACHE");
+	}
+#ifdef CONFIG_DCACHE_LINE_SIZE_DETECT
+	/* Dcache line size */
+	if (line_size <= 5) {
+		cache_cfg.data_line_size = 1 << (line_size + 2);
+	} else {
+		LOG_ERR("Unknown line size of D-cache");
+	}
+#elif (CONFIG_DCACHE_LINE_SIZE != 0)
+	cache_cfg.data_line_size = CONFIG_DCACHE_LINE_SIZE;
+#elif DT_NODE_HAS_PROP(DT_PATH(cpus, cpu_0), d_cache_line_size)
+	cache_cfg.data_line_size =
+		DT_PROP(DT_PATH(cpus, cpu_0), "d_cache_line_size");
+#else
+	LOG_ERR("Please specific the d-cache-line-size "
+		"CPU0 property of the DT");
+#endif /* CONFIG_DCACHE_LINE_SIZE_DETECT */
+#endif /* CONFIG_DCACHE */
+
+
+	if (!(csr_read(NDS_MMSC_CFG) & BIT(16))) {
+		LOG_ERR("Platform doesn't support I/D cache operation");
+	}
+
+#ifdef L2C_EXISTS
+	ret = nds_l2_cache_init();
+#endif /* L2C_EXISTS */
+
+	return ret;
+}
+
+SYS_INIT(andes_cache_init, PRE_KERNEL_1, CONFIG_APPLICATION_INIT_PRIORITY);
```

Comment:
CONFIG_APPLICATION_INIT_PRIORITY is not the right priority for initialization in PRE_KERNEL_1. Add a driver-specific priority or find an existing priority config appropriate for this driver.

Reply:
Thanks! I'll add a driver-specific priority.